pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+
+//! Oracle pallet to collect and aggregate oracle values.
+//!
+//! The collection admin configures the collection allowing a list of the oracle
+//! keys and from which feeders the values are collected.
+//!
+//! Later, the updating a collection will collect all values based on the admin
+//! configuration of the collection. The resulting collection is optimized to
+//! iterate though all values in jus one read.
+//!
+//! This pallet is not fed with external values, you need to configure a
+//! `ValueProvider` to provide it with oracle values.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{data::DataRegistry, PreConditions, ValueProvider};
+	use frame_support::{pallet_prelude::*, storage::bounded_btree_map::BoundedBTreeMap};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::{EnsureAddAssign, EnsureSubAssign, Zero};
+	use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+	use crate::{
+		types::{CachedCollection, KeyInfo},
+		util,
+	};
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type CollectionId: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// Represent the a time moment
+		type Moment: Parameter + Member + Copy + MaxEncodedLen + Ord;
+
+		/// A way to obtain oracle values from feeders
+		type OracleProvider: ValueProvider<
+			(Self::AccountId, Self::CollectionId),
+			Self::OracleKey,
+			Value = Self::OracleValue,
+			Timestamp = Self::Moment,
+		>;
+
+		/// Used to verify collection admin permissions
+		type IsAdmin: PreConditions<(Self::AccountId, Self::CollectionId), Result = bool>;
+
+		/// Max size of a data collection
+		#[pallet::constant]
+		type MaxCollectionSize: Get<u32>;
+
+		/// Max number of collections
+		#[pallet::constant]
+		type MaxFeedersPerKey: Get<u32>;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type Collection<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		BoundedBTreeMap<T::OracleKey, (T::OracleValue, T::Moment), T::MaxCollectionSize>,
+		ValueQuery,
+	>;
+
+	/// Store the keys that are registed for this collection
+	/// Only keys registered in this stored can be used to create the collection
+	#[pallet::storage]
+	pub(crate) type Keys<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::CollectionId,
+		Blake2_128Concat,
+		T::OracleKey,
+		KeyInfo<T>,
+		ValueQuery,
+	>;
```

Comment:
I'm quite happy because finally, it seems that we only need 2 storages in this pallet 😃